### PR TITLE
Early checking orderbook liquidity exhaustion

### PR DESCRIPTION
**Pull request by iboss-ptk**
- bump sqsdomain
- fix orderbook spot price miscalculation
- Revert "bump sqsdomain"
- extract oneBigDec to global space
- move shared logic into sqsdomain
- check exhaustion

## Summary by CodeRabbit

- **New Features**
  - Enhanced swap feasibility checks based on liquidity.
  - Improved spot price calculation with new logic.
  - Introduced new error messages for unsupported and duplicated denominators.

- **Bug Fixes**
  - Fixed issues related to direction determination and tick index retrieval.

- **Tests**
  - Added new test cases for orderbook direction, fillable amounts, and liquidity exhaustion calculations.

- **Chores**
  - Updated linting workflow with reordered steps and improved file matching patterns.

### CODE REVIEW

File: sqsdomain/cosmwasmpool/orderbook.go [hunk 1]
```diff
@@ -17,13 +17,95 @@ func (model *CosmWasmPoolModel) IsOrderbook() bool {
...
-	BaseDenom        string          `json:"base_denom"`
-	NextBidTickIndex int             `json:"next_bid_tick_index"` // tick index of the next bid tick, -1 if no bid tick
-	NextAskTickIndex int             `json:"next_ask_tick_index"` // tick index of the next ask tick, -1 if no ask tick
-	Ticks            []OrderbookTick `json:"ticks"`
+	QuoteDenom                     string          `json:"quote_denom"`
+	BaseDenom                      string          `json:"base_denom"`
+	NextBidTickIndex               int             `json:"next_bid_tick_index"`                 // tick index of the next bid tick, -1 if no bid tick
+	NextAskTickIndex               int             `json:"next_ask_tick_index"`                 // tick index of the next ask tick, -1 if no ask tick
+	BidAmountToExhaustAskLiquidity osmomath.BigDec `json:"bid_amount_to_exhaust_ask_liquidity"` // bid amount (in quote denom) to exhaust all ask liquidity
+	AskAmountToExhaustBidLiquidity osmomath.BigDec `json:"ask_amount_to_exhaust_bid_liquidity"` // ask amount (in base denom) to exhaust all bid liquidity
```
Comment (p0mvn) [inline sqsdomain/cosmwasmpool/orderbook.go]: Are these updated at time of ingest? If so, do you mind adding a short comment?

**Review by p0mvn — COMMENTED** [review]
Looks good so far!

Will wait for R4R

File: router/usecase/pools/routable_cw_orderbook_pool.go [hunk 2]
```diff
@@ -78,20 +80,29 @@ func (r *routableOrderbookPoolImpl) CalculateTokenOutByTokenIn(ctx context.Conte
...
+	if amountInRemaining.GT(amountInToExhaustLiquidity) {
+		return sdk.Coin{}, domain.OrderbookNotEnoughLiquidityToCompleteSwapError{PoolId: r.GetId(), AmountIn: tokenIn}
+	}
+
 	// ASSUMPTION: Ticks are ordered
 	for amountInRemaining.GT(zeroBigDec) {
-		// Order has run out of ticks to iterate
-		if tickIdx >= len(r.OrderbookData.Ticks) || tickIdx < 0 {
-			return sdk.Coin{}, domain.OrderbookNotEnoughLiquidityToCompleteSwapError{PoolId: r.GetId(), AmountIn: tokenIn}
-		}
```
Comment (p0mvn) [inline router/usecase/pools/routable_cw_orderbook_pool.go]: Should we still keep this check as defense-in-depth?

File: sqsdomain/cosmwasmpool/orderbook.go [hunk 3]
```diff
@@ -39,7 +122,65 @@ type OrderbookTickLiquidity struct {
...
 type OrderbookTick struct {
 	TickId        int64                  `json:"tick_id"`
 	TickLiquidity OrderbookTickLiquidity `json:"tick_liquidity"`
 }
+
+// CalcAmountInToExhaustOrderbookLiquidity calculates the amount of token in needed to exhaust all liquidity in the orderbook.
+// - orderDirection is the direction of the order to be placed and liquidity to be exhausted will be in the opposite direction.
+// - startingIndex is the index of the tick to start the calculation from.
+// - ticks is the list of ticks in the orderbook, assumed to be ordered by tick id in ascending order.
+func CalcAmountInToExhaustOrderbookLiquidity(directionIn OrderbookDirection, startingIndex int, ticks []OrderbookTick) (osmomath.BigDec, error) {
```
Comment (p0mvn) [inline sqsdomain/cosmwasmpool/orderbook.go]: Has this already been tested on the node together with the Osmosis side integration?
https://github.com/osmosis-labs/osmosis/pull/8458

File: router/usecase/pools/routable_cw_orderbook_pool.go (hunk 2 above)
Comment (iboss-ptk) [reply router/usecase/pools/routable_cw_orderbook_pool.go]: sure thing

File: sqsdomain/cosmwasmpool/orderbook.go (hunk 3 above)
Comment (iboss-ptk) [reply sqsdomain/cosmwasmpool/orderbook.go]: yes

**Review by p0mvn — APPROVED** [review]

**Pull request merged**